### ISSUE TITLE
Causality issue with compressive attention

### ISSUE THREAD

**Issue by mutiann** [tilde-research/nsa-impl#3]
Hi,

Thank you for providing the code, and the observations in the blogpost are really insightful! While I noticed a possible issue in the compressive attention implementation, as

https://github.com/tilde-research/nsa-impl/blob/623e2c5a30a6979e1fc0cf264c09b66963b1e175/nsa/nsa.py#L40-L41

, which appears to be a causal mask to prevent attending to future compressed blocks. While according to my understanding to a casual mask in FlexAttention, the mask function should return True if the position should be included (i.e. not masked), as we can find in the official example:
```
from torch.nn.attention.flex_attention import create_block_mask

def causal(b, h, q_idx, kv_idx):
    return q_idx >= kv_idx

# Because the sparsity pattern is independent of batch and heads, we'll set them to None (which broadcasts them) 
block_mask = create_block_mask(causal, B=None, H=None, Q_LEN=1024, KV_LEN=1024)
```

Hence I'm a bit confused and wondering if the masking is correctly implemented. Thank you!

**Comment by Houxin2024** [comment]
Confirming this in end-to-end training.

  We vendored `623e2c5` into a 1.2B MoE run (GQA, num_query_groups=1, seq=8192, GBS=1024, bf16).
  Training behavior:
  - step 0–200: tracks our dense baseline closely (~1.2 CE)
  - step ~200 onward: loss smoothly decays to ~0 (PPL ≈ 1), both train and val
  - Textbook causal-leak signature — model progressively learns to exploit the leaked branch.

I think fix:
   def cmp_mask(b, h, q_idx, kv_idx):
  -    return q_idx <= (kv_idx + 1) * block_size - 1
  +    return q_idx >= (kv_idx + 1) * block_size
